tests/how_to_test_commands.rs

```diff
@@ -0,0 +1,54 @@
+use bevy::ecs::system::Command;
+use bevy::prelude::*;
+
+struct Request;
+
+#[derive(Component)]
+struct Requests(Vec<Request>);
+
+impl Requests {
+    fn new(request: Request) -> Self {
+        Self([request].into())
+    }
+
+    fn add(&mut self, request: Request) {
+        self.0.push(request);
+    }
+
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+struct InsertOrAddRequest(pub Entity, pub Request);
+
+impl Command for InsertOrAddRequest {
+    fn write(self, world: &mut World) {
+        let Self(entity, request) = self;
+        if let Some(mut requests) = world.get_mut::<Requests>(entity) {
+            requests.add(request);
+        } else {
+            world.entity_mut(entity).insert(Requests::new(request));
+        }
+    }
+}
+
+#[test]
+fn add_request() {
+    use bevy::ecs::system::ApplyCommands;
+
+    let mut world = World::new();
+    let entity = world.spawn_empty().id();
+    world.apply_commands(|_, mut commands| {
+        commands.add(InsertOrAddRequest(entity, Request));
+        commands.add(InsertOrAddRequest(entity, Request));
+    });
```

Comment:
Why is this preferable to the existing:

```rust
InsertOrAddRequest(entity, Request).write(&mut world);
InsertOrAddRequest(entity, Request).write(&mut world);
```

It doesn't add a new abstraction, uses fewer lines of code, and is more efficient (because it doesn't use an intermediate `Commands` buffer).

Reply:
It's often desirable to write utility functions that take a `&mut Commands` in game logic. For example:
```rust
fn spawn(commands: &mut Commands) -> Entity {
  let bundle = todo!("Make a bundle");
  commands.spawn(bundle).id()
}
```
Without this changelist, I can't test such a function without manually creating a `CommandQueue`, or a duplicate version of the function which works on `&mut World`.

Reply:
Fair point! I agree that Commands access might be necessary in some cases. But this should _only_ really be used for tests as allocating and immediately dropping `CommandBuffer` isn't a pattern we should be encouraging.

Is the risk of making this easy really worth it? Especially when this is almost as ergonomic?
```rust
let mut schedule = Schedule::new();
schedule
    .add_systems(move |mut commands: Commands| {
        commands.add(InsertOrAddRequest(entity, Request));
        commands.add(InsertOrAddRequest(entity, Request));
    })
    .run(&mut world);
```

And if we reeeeeealllly want to make this ergonomic, maybe we should just do this:

```rust
world.run_system(move |mut commands: Commands| {
    commands.add(InsertOrAddRequest(entity, Request));
    commands.add(InsertOrAddRequest(entity, Request));
});
```

Also an anti-pattern in game code (because it will create the system state and immediately drop it). But it removes the special casing of commands and makes this more generally useful.

Reply:
I would agree that `run_system` is a superior alternative to `apply_commands`.

Let me know if you'd like me to modify this CL into `run_system`.

Reply:
On second thought, there is a downside to `run_system` compared to `apply_commands`.

`apply_commands` returns the return value of its function. This allows the function to return the list of entities spawned, for example. As far as I know, we wouldn't be able to get the output of a system with `run_system`. Right...?

Reply:
We could hook up `run_system` to the system piping TBH.

Reply:
If `run_system` could return the output of a piped system, that would be most ideal. I'm not sure how I'd approach implementing that though. I could give it a try with some guidance/examples though.

Reply:
> apply_commands returns the return value of its function. This allows the function to return the list of entities spawned, for example. As far as I know, we wouldn't be able to get the output of a system with run_system. Right...?

System impls can have outputs (return values) and run_system can return them. No need for any new abstractions. It would "just work".

Reply:
I do think we should port this to run_system.

Reply:
Sounds good to me. I'll close this PR and open a new one for `run_system` once I have a working prototype.